src/Controllers/FilesController.js

```diff
@@ -42,22 +44,167 @@ export class FilesController extends AdaptableController {
+    const referencesToAdd = [];
+    const referencesToRemove = [];
+    const searchForSubfiles = (keyData, keyOriginalData) => {
+      if (typeof keyData !== 'object') {
+        return;
+      }
+      if (!keyOriginalData) {
+        keyOriginalData = {};
+      }
+      for (const key in keyData) {
+        const val = keyData[key] || {};
+        const original = keyOriginalData[key] || {};
+        if (typeof val !== 'object') {
+          continue;
+        }
+        const { __type, name } = val;
+        if (__type === 'File' || original.__type == 'File') {
+          if (name === original.name) {
+            continue;
+          }
+          if (name && original.name == null) {
+            referencesToAdd.push(name);
+          }
+          if (original.name && name == null) {
+            referencesToRemove.push(name);
+          }
+          continue;
+        }
+        searchForSubfiles(val, original);
+      }
+    };
+    searchForSubfiles(data, originalData);
+    const allFiles = referencesToAdd.concat(referencesToRemove);
+    if (allFiles.length == 0) {
+      return;
+    }
+    const filesToFind = allFiles.map(val => new Parse.File(val).toJSON());
+    const fileQuery = new Parse.Query('_File');
+    fileQuery.containedIn('file', filesToFind);
+    const fileData = await fileQuery.find({ useMasterKey: true });
+    const filesToSave = [];
+    for (const fileObject of fileData) {
+      const { _name } = fileObject.get('file');
+      if (referencesToAdd.includes(_name)) {
+        fileObject.addUnique('references', {
+          objectId: data.objectId,
+          className,
+        });
+      } else {
+        fileObject.remove('references', { objectId: data.objectId, className });
+      }
+      filesToSave.push(fileObject);
+    }
+    await Parse.Object.saveAll(filesToSave, { useMasterKey: true });
+  }
+  async getAuthForFile(config, file, auth, object, className) {
+    if (className === '_File') {
+      return;
+    }
+    const [fileObject] = await config.database.find('_File', {
+      file,
+    });
+    if (!fileObject) {
+      return;
+    }
+    let toSave = false;
+    const tokens = fileObject.tokens || [];
+    const allowed = await this.canViewFile(config, fileObject, auth.user);
+    if (allowed) {
+      const url = file.url.split('?');
+      let appendToken = '';
+      if (url.length > 1) {
+        appendToken = `${url[1]}&`;
+      }
+      const token = randomString(25);
+      appendToken += `token=${token}`;
+      file.url = `${url[0]}?${appendToken}`;
+      const expiry = new Date(new Date().getTime() + 30 * 60000);
+      tokens.push({
+        token,
+        expiry,
+        user: auth.user,
+      });
+      toSave = true;
+    }
+    const references = fileObject.references || [];
+    if (!references.includes({ objectId: object.objectId, className })) {
+      references.push({ objectId: object.objectId, className });
+      toSave = true;
+    }
```

Comment:
If i understand, reference is an Array of Pointer. Here it could be a bottle neck if a file has too many references :) 
(Like a admin upload an avatar, and all the user base use it). May be reference could be only a simple native DB counter (transaction activated) to keep how many objects use it.

Reply:
Also references update should be ACID, otherwise we could start to have invalid `reference` count/array.

Reply:
I have updated this on the latest revision, pls let me know if I understood you correctly. Thank you so much for you feedback